Add CoRIM extension mechanisms (#96)
veraison/corim: extensions/README.md

[CoRIM
specification](https://www.ietf.org/archive/id/draft-ietf-rats-corim-02.htm)

> may be extended by other specifications at several points identified in the

Reviewer: ```suggestion
may be extended by other specifications at well defined points identified in the
```

spec using CDDL extension sockets. This implementation, likewise, allows
dependent code to register extension types. This is done via three distinct
extension mechanisms:

1. Some structures embed an Extensions object that allows registering a
   user-provided struct. The fields of that struct will be treated as extension
   fields to embedding structure. This corresponds to the map-extension sockets
   in the spec.
2. Some elements can contain values of one of pre-determined types. New types
   can be provided by registering a factory function for those types using an
   appropriate function. This (mostly) corresponds to the type-choice extension
   sockets in the spec.
3. A couple of type-choice sockets (`$tag-rel-type-choice`,
   `$corim-role-type-choice` and `$comid-role-type-choice`) define what, in
   effect, are extensible enums. They allow providing additional values, rather
   than types. This implementation provides registration functions for new
   values for those types.

> 

Author: I have reworded to hopefully make thigns clearer, however I kept the numbed bullet formatting as this a list of the three extension mechanisms mentioned in the preceeding paragraph and ought to be formatted as such.

> [!NOTE]
> The CoRIM spec includes some CDDL from the CoSWID spec that also features
> extension sockets. This code base does include the CoSWID implementation, and

> > so does not cover those extension points.

Author: Reworded.

> ## Map Extensions

Reviewer: So Far we have been using the term `structures` for extending CoRIM Maps

Author: Theere are no structures in CoRIM, in the problem domain/concemptually, they are CoRIM map extensions, that are implemented suing `Extensions` `structs`. "Map" and "type choice" are the names of the extensions types being implemented.

introducing additional constraints over existing fields.

### Adding new fields

To add a new fields, simply add them to your extensions struct, ensuring that
the `cbor` and `json` tags on those fields are set corrects. As CoRIM spec
mandates integer keys, you must use `keyasint` option for the `cbor` tag.

To access the values of those fields, you can call the extended type instance's
`Extensions.Get()` passing in the name of the field you want to access. The
name can be either the Go struct's field name, the name specified in the `json`
tag, or (a string containing) the integer specified in the `cbor` tag.

`Get()` returns an `interface{}`. There are equivalent `GetInt()`,
`GetString()`, etc. methods that  perform the required conversions, and return
the value of the indicated type, along with possible errors. ("Must" versions
of these also exists, e.g. `MustGetString()`, that do not return an error).

You can also get the pointer your extensions instance itself by calling
extended type instance's `GetExtensions()`. This returns an `interface{}`, so
you will need to type assert in order to be able to access the fields directly.

> ### Introducing additional constraints

Reviewer: ```suggestion
#### Introducing additional constraints
```

Author: no three is correct -- it should be under "Map Extensions", at the same level as "adding new fields" as those are the two things you can do with map extensions.

	_, err := uuid.Parse(val.EntityName.String())
	if err != nil {
		return fmt.Errorf("invalid UUID: %w", err)
	}

	return nil
}

var sampleText = `
{
      "name": "31fb5abf-023e-4992-aa4e-95f9c1503bfa",
      "regid": "https://acme.example",
      "email": "info@acme.com",
      "roles": [
        "tagCreator",
        "creator",
        "maintainer"
      ]
}
`

func main() {
	var entity comid.Entity
	entity.RegisterExtensions(&EntityExtensions{})

	if err := json.Unmarshal([]byte(sampleText), &entity); err != nil {
		log.Fatalf("ERROR: %s", err.Error())
	}

	if err := entity.Valid(); err != nil {
		log.Fatalf("failed to validate: %s", err.Error())
	} else {
		fmt.Println("validation succeeded")
	}

    // obtain the extension field value via a generic getter
	email := entity.Extensions.MustGetString("email")
	fmt.Printf("entity email: %s\n", email)

    // retrive the extensions struct and get value via its field.
	exts := entity.GetExtensions().(*EntityExtensions)
	fmt.Printf("also entity email: %s\n", exts.Email)
}
```

## Type Choice Extensions

Type Choice extensions allow specifying alternative types for existing CoRIM

> fields by defining a struct that implements an appropriate interface and

Reviewer: ```suggestion
fields by defining a  type that implements an appropriate interface and
```

Reviewer: I am confused, I thought for Type extensions we will register the new type and a Callback which instantiates the new Type.

Why we are requesting users to create a structure?

Author: - "struct" rather than "type" is correct here, since we're talking about concrete implementations (the interface being implemented is the type).
  - I assume, by "callback" you mean "factory"? The factory still needs something to actually put into the new instance that that something is the struct that needs to be implemented. E.g. need to provide an implementation for both `IClassIDValue` which is what you're actually implementing, and a `IClassIDFactory` that produces a `*ClassID` that is populated with the underlying value of your type.

Reviewer: > * "struct" rather than "type" is correct here, since we're talking about concrete implementations (the interface being implemented is the type).
> * I assume, by "callback" you mean "factory"? The factory still needs something to actually put into the new instance that that something is the struct that needs to be implemented. E.g. need to provide an implementation for both `IClassIDValue` which is what you're actually implementing, and a `IClassIDFactory` that produces a `*ClassID` that is populated with the underlying value of your type.

As discussed the user can define a type like taggedInt and provide the required methods for instantiation.

I think, we need to refine the language, to make it super clear to the consumer of the extensions!

base CoRIM spec using CDDL extension sockets. This implementation, likewise, allows
dependent code to register extension types. This is done via three distinct
extension mechanisms:

1. Structures that allow extensions embed an `Extensions` object, which allows
   registering a user-provided struct. The fields of that struct will be
   treated as extension fields to embedding structure. This corresponds to the
   map-extension sockets in the spec.
2. Some fields may have values one of a set of pre-determined types (e.g. an
   instance ID may be either a UUID or UEID). These (mostly) correspond to
   type-choice sockets in the CoRIM spec. The set of allowed types for a field
   may be extended by registering a factory function for a new type, using the
   registration function associated with the type choice.
3. A couple of type-choice sockets (`$tag-rel-type-choice`,
   `$corim-role-type-choice` and `$comid-role-type-choice`) define what, in
   effect, are extensible enums. They allow providing additional values, rather
   than types. This implementation provides registration functions for new
   values for those types.

> > [!NOTE]
> > Some of the definitions in the CoRIM spec are "imported" from the CoSWID
> > spec.  Some of those definitions also feature extension sockets. However, as
> > they are defined in a different spec (and are not implemented in this
> > library), they cannot be extended using the above mechanisms.
> 

Reviewer: ```suggestion
> [!NOTE]
> CoRIM specification also "imports" CDDL from the CoSWID spec. Some of 
> these COSWID CDDL definitions also feature extension sockets. 

> However, as they are defined in a different spec and are implemented 
> in `veraison/coswid` library, they cannot be extended 
> using the extension feature provided in the CoRIM Library. The extension 
> support in CoRIM library is applicable ONLY to CoRIM  and CoMID Maps and 
> Types choices
```

Reviewer: Request please review an attempt to clarify the Note point.

Feel free to edit my text to make it more clear, if you feel so!

## Map Extensions

Map extensions allow extending CoRIM maps with additional keys, effectively
defining new fields for the corresponding structures. In the code base, these
can be identified by the embedded `Extensions` struct. These are

- `comid.Comid`
- `comid.Entity`
- `comid.FlagsMap`
- `comid.Mval`
- `comid.Triples`
- `corim.Entity`
- `corim.Signer`
- `corim.UnsignedCorim`

> 

Reviewer: Please make note that the list will change as the specification evolves, this is applicable to `draft-02` of the specification

Author: There is a note in the top-level README to that effect. We should not be idetifying specific things that may change (because the implication then becomes that unless something is expressly indicated as subject to change, it won't change).

Reviewer: Thanks, so long as we have note it is fine!

To extend the above types, you need to define a struct containing your
extensions and pass a pointer to an instance of that struct to the
`RegisterExtensions()` method of the corresponding instance of the type that is
being extended. This should be done as early as possible, before any marshaling
is performed.

These types can be extended in two ways: by adding additional fields, and by
introducing additional constraints over existing fields.

### Adding new fields

To add a new fields, simply add them to your extensions struct, ensuring that

> the `cbor` and `json` tags on those fields are set corrects. As CoRIM spec

Reviewer: ```suggestion
the `cbor` and `json` tags on those fields are set correctly. As CoRIM spec
```

name can be either the Go struct's field name, the name specified in the `json`
tag, or (a string containing) the integer specified in the `cbor` tag.

`Get()` returns an `interface{}`. There are equivalent `GetInt()`,
`GetString()`, etc. methods that  perform the required conversions, and return
the value of the indicated type, along with possible errors. ("Must" versions
of these also exists, e.g. `MustGetString()`, that do not return an error).

You can also get the pointer your extensions instance itself by calling
extended type instance's `GetExtensions()`. This returns an `interface{}`, so
you will need to type assert in order to be able to access the fields directly.

### Introducing additional constraints

To introduce new constraints,  add a method called `Validate<type>(v *<type>)`
to your extensions struct, where `<type>` is the name of the type being
extended (one of the ones listed above) -- e.g.
`ValidateComid(v *comid.Comid)` when extending `comid.Comid`. This method, if
it exists, will be invoked inside the extended type instance's `Valid()`
method, passing itself as the parameter.

You do not need to define this method, unless you actually want to enforce some

> constraints (i.e. if you just want to define additional fields).

Reviewer: ```suggestion
constraints.
```

Author: Why remove this clarification?

Reviewer: The text is confusing, are you providing additional information in the braces `( )` ?

I found it to be counter intuitive to your original statement, which says !
Perhaps may be we simply say:
You do not need to define this method, if  additional constraints are not introduced by the new fields.

Author: >if additional constraints are not introduced by the new fields.

??? the two things are orthogonal:

- new fields (do not require validation method -- constraints on new fields can be enforced via the fields' types)
- additional constraints (either on existing fields, or inter-field constraints)

The parentheciesd statement clarifies the situations where you might not want to add additional contraitas (when you're only adding new fields).

Reviewer: Thanks for all the descriptions. I think, a clarification is needed at the top to describe what this extension implies.

Then this explaination can have the right context

### Example

The following example illustrates how to implement a map extension by extending `comid.Entity` with the following features:

1. an optional "email" field

> 2. additional validation to ensure that the existing name field contains a

Reviewer: I have a clarification question here more than a comment.

What is the `use case` you are trying to highlight where the Validate method 
is checking for existing element validity ?

My concern is that a new user of extensions will get further confused on how the Validate is useful ?

For the benefit of simplicity can we keep it just checking the new parameters added for validity (like URI added is a valid URI?)

Author: You _can_ use Validate*  to  enforce individual field constraits on new fields, but it is not necessary for that, and it is not the reason it exists.  It is not necessary for that.

The things that the Valadiate* methods make possible that would not be possible other wise are:

- validating additional contraints on existing fields (since the types of those fields are fixed, you cannot do it the usual way, like you can with new fields)
- valdiating inter-field constraints (which, by their nature, cannot be captured in a type) 

The example illustrates that.

The primary mechanism for enforcing constraints on individual values is types. that's the sole reason for having types in the first place.

> OR I am setting an new Identifier which can be []byte but valid ONLY
when size is fixed to 32 bytes or 64 bytes.

You can do this by introducing your won type.  which is defined as "a seequence of exactly 32 or 64 bytes". 

If the example only showed validation of a new field, _that_ would be a sourc of confusion -- "why do I need to define Vadate method that take a whole other struct as an object, when I can just define my own type?"

`RegisterExtensions()` method of the corresponding instance of the type that is
being extended. This should be done as early as possible, before any marshaling
is performed.

These types can be extended in two ways: by adding additional fields, and by
introducing additional constraints over existing fields.

### Adding new fields

To add a new fields, simply add them to your extensions struct, ensuring that
the `cbor` and `json` tags on those fields are set correctly. As CoRIM spec
mandates integer keys, you must use `keyasint` option for the `cbor` tag.

To access the values of those fields, you can call the extended type instance's
`Extensions.Get()` passing in the name of the field you want to access. The
name can be either the Go struct's field name, the name specified in the `json`
tag, or (a string containing) the integer specified in the `cbor` tag.

`Get()` returns an `interface{}`. There are equivalent `GetInt()`,
`GetString()`, etc. methods that  perform the required conversions, and return
the value of the indicated type, along with possible errors. ("Must" versions
of these also exists, e.g. `MustGetString()`, that do not return an error).

> You can also get the pointer your extensions instance itself by calling

Reviewer: ```suggestion
You can also get the pointer to your extensions instance itself by calling
```

> [CoRIM
> specification](https://www.ietf.org/archive/id/draft-ietf-rats-corim-02.htm)

Reviewer 2: ```suggestion
The [CoRIM
data format](https://datatracker.ietf.org/doc/draft-ietf-rats-corim)
```

Author: I'm specificicallying linking to draft 2 here since that is what we're following; The latest version may not match.

Reviewer 2: OK.  My comment was about using the IETF datatracker's which is supposed to have stronger URL stability.  If we need a precise version we use the versioned link https://datatracker.ietf.org/doc/draft-ietf-rats-corim/02

may be extended by CoRIM Profiles documented in other specifications at well
defined points identified in the base CoRIM spec using CDDL extension sockets.
CoRIM profiles may

1. Introduce new data fields to certain objects.
2. Further constrain allowed values for existing fields.

> 3. Introduce new type choices for fields who's values can be one of several

Reviewer 2: ```suggestion
3. Introduce new type choices for fields whose values can be one of several
```

   types.

This implementation, likewise, allows dependent code to register extension
types. This is done via three distinct extension mechanisms:

1. Structures that allow extensions embed an `Extensions` object, which allows
   registering a user-provided struct. The user-provided struct can extend
   their containing structures in two ways:

   - the fields of the user-provided struct become additional fields in
     the containing structure.

>    - the user-provided struct may define addition constraints on the containing

Reviewer: ```suggestion
   - the user-provided struct may define additional constraints on the containing
```

     containing structure by defining an appropriate validation method for it.

   This corresponds to the map-extension sockets in the spec.

> 2. Some fields may have values one of a set of pre-determined types (e.g. an

Reviewer 2: ```suggestion
2. Some fields may have values chosen from a set of pre-determined types (e.g., an
```

   instance ID may be either a UUID or UEID). These (mostly) correspond to
   type-choice sockets in the CoRIM spec. The set of allowed types for a field
   may be extended by registering a factory function for a new type, using the
   registration function associated with the type choice.
3. A couple of type-choice sockets (`$tag-rel-type-choice`,
   `$corim-role-type-choice` and `$comid-role-type-choice`) define what, in
   effect, are extensible enums. They allow providing additional values, rather
   than types. This implementation provides registration functions for new
   values for those types.

> [!NOTE]

> > CoRIM specification also "imports" CDDL from the CoSWID spec. Some of
> > these COSWID CDDL definitions also feature extension sockets.
> 
> > However, as they are defined in a different spec and are implemented
> > in `veraison/coswid` library, they cannot be extended
> > using the extension feature provided in the CoRIM Library. The extension
> > support in CoRIM library is applicable ONLY to CoRIM  and CoMID Maps and
> > Types choices

Reviewer 2: ```suggestion
> CoRIM also "imports" CDDL from the CoSWID spec. Some of
> these CoSWID CDDL definitions also feature extension sockets.
> However, as they are defined in a different spec and are implemented
> in the [`veraison/swid`](https://github.com/veraison/swid) library, they cannot be extended
> using the extension feature provided in the CoRIM library. The extension
> support in CoRIM library is applicable ONLY to CoRIM and CoMID maps and
> type choices
```

## Map Extensions

Map extensions allow extending CoRIM maps with additional keys, effectively
defining new fields for the corresponding structures. In the code base, these
can be identified by the embedded `Extensions` struct. These are

- `comid.Comid`
- `comid.Entity`
- `comid.FlagsMap`
- `comid.Mval`
- `comid.Triples`
- `corim.Entity`
- `corim.Signer`
- `corim.UnsignedCorim`

To extend the above types, you need to define a struct containing your
extensions and pass a pointer to an instance of that struct to the
`RegisterExtensions()` method of the corresponding instance of the type that is
being extended. This should be done as early as possible, before any marshaling
is performed.

These types can be extended in two ways: by adding additional fields, and by
introducing additional constraints over existing fields.

### Adding new fields

> To add a new fields, simply add them to your extensions struct, ensuring that
> the `cbor` and `json` tags on those fields are set correctly. As CoRIM spec
> mandates integer keys, you must use `keyasint` option for the `cbor` tag.

Reviewer 2: ```suggestion
To add new fields, simply add them to your extensions struct, ensuring that
the `cbor` and `json` tags on those fields are set correctly. As CoRIM
mandates integer keys, you must use the `keyasint` option for the `cbor` tag.
```

> To access the values of those fields, you can call the extended type instance's
> `Extensions.Get()` passing in the name of the field you want to access. The
> name can be either the Go struct's field name, the name specified in the `json`
> tag, or (a string containing) the integer specified in the `cbor` tag.
> 
> `Get()` returns an `interface{}`. There are equivalent `GetInt()`,
> `GetString()`, etc. methods that  perform the required conversions, and return
> the value of the indicated type, along with possible errors. ("Must" versions
> of these also exists, e.g. `MustGetString()`, that do not return an error).

Reviewer 2: ```suggestion
To access the values of those fields, you can call the extended type instance's
`Extensions.Get()` passing in the name of the field you want to access. The
name can be either the Go struct's field name, the name specified in the `json`
tag, or (a string containing) the integer specified in the `cbor` tag.

`Get()` returns an `interface{}`. There are equivalent `GetInt()`,
`GetString()`, etc. methods that perform the required conversions, and return
the value of the indicated type, along with possible errors. ("Must" versions
of these also exist, e.g. `MustGetString()`, that do not return an error and instead call `panic()`).
```

> You can also get the pointer your extensions instance itself by calling
> extended type instance's `GetExtensions()`. This returns an `interface{}`, so
> you will need to type assert in order to be able to access the fields directly.

Reviewer 2: ```suggestion
You can also get the pointer to your extension's instance itself by calling
the extended type instance's `GetExtensions()`. This returns an `interface{}`, so
you will need to type assert to be able to access the fields directly.
```

### Introducing additional constraints

> To introduce new constraints,  add a method called `Valid<TYPE>(v *<TYPE>)`

Reviewer: I am thinking, really we are restricting and Constraining the type, here.

Should we name the method as 

`Constrain  <TYPE>' ?

Reviewer: I mean ConstrainType

Author: done.

> To introduce new constraints,  add a method called `Constrain<TYPE>(v *<TYPE>)`
> to your extensions struct, where `<TYPE>` is the name of the type being
> extended (one of the ones listed above) -- e.g.
> `ConstrainComid(v *comid.Comid)` when extending `comid.Comid`. This method, if
> it exists, will be invoked inside the extended type instance's `Valid()`
> method, passing itself as the parameter.

Reviewer 2: ```suggestion
To introduce new constraints, add a method called `Constrain<TYPE>(v *<TYPE>)`
to your extensions struct, where `<TYPE>` is the name of the type being
extended (one of the ones listed above) -- e.g.
`ConstrainComid(v *comid.Comid)` when extending `comid.Comid`. This method, if
it exists, will be invoked inside the extended type instance's `Valid()`
method, passing itself as the parameter.
```

> You do not need to define this method, unless you actually want to enforce some
> constraints (i.e. if you just want to define additional fields).

Reviewer 2: ```suggestion
You do not need to define this method unless you actually want to enforce some
constraints (i.e., if you just want to define additional fields).
```

### Example

The following example illustrates how to implement a map extension by extending `comid.Entity` with the following features:

1. an optional "email" field

> 2. additional validation to ensure that the existing name field contains a

Reviewer: ```suggestion
2. additional `constrain method` to ensure that the existing name field contains a
```

}

// custom constraints for Entity
func (o EntityExtensions) ConstrainEntity(val *comid.Entity) error {
	_, err := uuid.Parse(val.EntityName.String())
	if err != nil {
		return fmt.Errorf("invalid UUID: %w", err)
	}

	return nil
}

var sampleText = `
{
      "name": "31fb5abf-023e-4992-aa4e-95f9c1503bfa",
      "regid": "https://acme.example",
      "email": "info@acme.com",
      "roles": [
        "tagCreator",
        "creator",
        "maintainer"
      ]
}
`

func main() {
	var entity comid.Entity
	entity.RegisterExtensions(&EntityExtensions{})

	if err := json.Unmarshal([]byte(sampleText), &entity); err != nil {
		log.Fatalf("ERROR: %s", err.Error())
	}

	if err := entity.Valid(); err != nil {
		log.Fatalf("failed to validate: %s", err.Error())
	} else {
		fmt.Println("validation succeeded")
	}

    // obtain the extension field value via a generic getter
	email := entity.Extensions.MustGetString("email")
	fmt.Printf("entity email: %s\n", email)

    // retrive the extensions struct and get value via its field.
	exts := entity.GetExtensions().(*EntityExtensions)
	fmt.Printf("also entity email: %s\n", exts.Email)
}
```

## Type Choice Extensions

> Type Choice extensions allow specifying alternative types for existing CoRIM
> fields by defining a type that implements an appropriate interface and
> registering it with CBOR tag.
> 
> A type choice struct contains a single field, `Value`, that contains the actual
> object represented by the type choice. The `Value` implements an interface
> that is specific to the type choice and is derived from `ITypeChoiceValue`:

Reviewer 2: ```suggestion
Type Choice extensions allow specifying alternative types for existing CoRIM
fields by defining a type that implements an appropriate interface and
registering it with a CBOR tag.

A type choice struct contains a single field, `Value`, that contains the actual
object represented by the type choice. The `Value` implements an interface
that is specific to the type choice and is derived from `ITypeChoiceValue`:
```

```go
type ITypeChoiceValue interface {
	// String returns the string representation of the ITypeChoiceValue.
	String() string
	// Valid returns an error if validation of the ITypeChoiceValue fails,
	// or nil if it succeeds.
	Valid() error
	// Type returns the type name of this ITypeChoiceValue implementation.
	Type() string
}
```

The following is the full list of type choice structs:

- `comid.ClassID`
- `comid.CryptoKey`
- `comid.EntityName`
- `comid.Group`
- `comid.Instance`
- `comid.Mkey`
- `comid.SVN`
- `corim.EntityName`

To provide a new value type, the following is required:

> 1. Define a type that implements the value interface for the type choice you
>    want to extend. This interface is called `I<NAME>Value`,  where `<NAME>` is
>    the name of the type choice type(e.g. `IClassIDValue`).  These interfaces
>    always embed `ITypeChoiceValue` and possible define additional methods.

Reviewer 2: ```suggestion
1. Define a type that implements the value interface for the type choice you
   want to extend. This interface is called `I<NAME>Value`,  where `<NAME>` is
   the name of the type choice type(e.g. `IClassIDValue`).  These interfaces
   always embed `ITypeChoiceValue` and possibly define additional methods.
```

2. Create a factory function for your type, with the signature `func (any)
   (*<NAME>, error)`, where `<NAME>` is the name of the type choice type that
   will contain your value. (Note that the function must return a pointer to
   the container type choice struct, _not_ to the value type you define.) This
   function should create an instance of your value type from the provided
   input and return a new type choice struct instance containing it. The range
   of valid inputs is up to you, however it _must_ handle `null`, returning the
   [zero-value](https://go.dev/ref/spec#The_zero_value) creating the zero value

>    for your type in that case.

Reviewer 2: _"[...] however it _must_ handle `null`, returning the [zero-value](https://go.dev/ref/spec#The_zero_value) creating the zero value for your type in that case."_

There seems to be some missing punctuation, or some amount of redundancy, which makes it difficult to understand.

Author: urgh, yeah, the "creating the zero value" bit is redundant.

3. Register your factory function with the CBOR tag for your new type by
   passing it to the registration function corresponding to the type choice
   struct. It will have the name `Register<NAME>Type`, where `<NAME>` is the
   name of the type choice struct that will contain your value (e.g.
   `RegisterClassIDType`).

### Example

The following example illustrates how to add a new type choice value

> implementation by extending `CryptoKey` type to support DER values.

Reviewer 2: ```suggestion
implementation by extending the `CryptoKey` type to support DER values.
```
